expat/xmlwf/xmlwf.c

```diff
@@ -1061,52 +1075,55 @@ tmain(int argc, XML_Char **argv) {
       userData.fp = tfopen(outName, T("wb"));
       if (! userData.fp) {
         tperror(outName);
-        exit(3);
-      }
-      setvbuf(userData.fp, NULL, _IOFBF, 16384);
+	if (!continueOnError) {
+	  exit(3);
+	}
```

Comment:
We need to record exit code 3 to variable `exitCode` so that it doesn't get lost — no? Right now the pull request seems to make xmlwf exit with status 0 rather than 3 for these errors.

This alternative code seems to do that and save the whole re-indentation diff trouble:
```c
exitCode = 3;
if (continueOnError) {
  free(outName);
  continue;
} else {
  break;
}
```
What do you think?

If we *do* continue with the re-indentation approach, we'll need to get the leading tab characters out of there.
I'm making the CI detect leading tabs in #440 now so that there is no chance to have leading tabs go unnoticed again.

PS: When we encounter exit status 2 _and_ 3 during a run, currently the last would win. If that's not in your interest, we can make a new status code say 5 that means we have seen 2 and 3 — just an idea.

Reply:
You're right on the code getting lost in the case that -k is specified.  Oops, I'll set exitCode.

Here's the thinking on the big indented "else" clause.  When opening the output file fails, the PR still goes ahead and processes the file, potentially reporting XML errors.  But to do that, it's necessary to bypass all the code in that else, which does a bunch of stuff with userData.fp to set up the output. 

If we said that if the output file can't be opened then we'll just skip parsing the file altogether, then we can use your continue/break approach and get rid of the big else.  I slightly but not strongly prefer the current PR approach, but if you have a strong opinion I'm not fussy.

As for the case where you hit both 2 and 3, exit code 5 seems plausible.  Easy change. But… might be better to get this PR squared away and then I'll do another one to clean up the exit codes, also move the current 2 for arglist errors to 4.  A separate PR might be nice because maybe someone out there in the community has a strong dependence on exit codes and this gives them a chance to freak out?

Oops, pardon me on the tabs, I'll make sure to untabify in future.  Been a while since I did any C coding.

Also OK with your other points on indentation and language, will fix.

Reply:
> Here's the thinking on the big indented "else" clause. When opening the output file fails, the PR still goes ahead and processes the file, potentially reporting XML errors. But to do that, it's necessary to bypass all the code in that else, which does a bunch of stuff with userData.fp to set up the output.
> 
> If we said that if the output file can't be opened then we'll just skip parsing the file altogether, then we can use your continue/break approach and get rid of the big else. I slightly but not strongly prefer the current PR approach, but if you have a strong opinion I'm not fussy.

Writing to file needs argument `-d DIRECTORY` to happen. xmlwf will write (a) canonical XML or (b) "meta XML" or (c) a copy of the input then. Well-formedness is a hard requirement to (a) and (b) at least and I would argue that in those cases the user is after the output, not after well-formedness.
For (c) I'm not too sure about the main point of that mode frankly: It will need the upcoming `-k` to work as a "copy all well-formed XML files among the given ones" which then has some point I guess… but before that? Given that the output of xmlwf is not made to be machine readable… is there value to still parse those files if they cannot be copied and the exit status will be non-zero anway? What if _each and all of them_ fail to write (e.g. because the target directory is not writable) and then we parsed all those files "for nothing" and could have returned way quicker? I don't know, just thinking aloud.

> As for the case where you hit both 2 and 3, exit code 5 seems plausible. Easy change. But… might be better to get this PR squared away and then I'll do another one to clean up the exit codes, also move the current 2 for arglist errors to 4. A separate PR might be nice because maybe someone out there in the community has a strong dependence on exit codes and this gives them a chance to freak out?

A separate PR after this one with exit codes 4 and 5 sounds good!


> Oops, pardon me on the tabs, I'll make sure to untabify in future. Been a while since I did any C coding.

Maybe it's time to add a [`.editorconfig` file] to this repository. Does your editor support [EditorConfig](https://editorconfig.org/)?